### PR TITLE
Silence some complaints from valgrind

### PR DESCRIPTION
**Pull request by rincebrain**
### Motivation and Context
Without this, valgrind is effectively useless in most cases, as
it floods with messages about zio writes trying to read uninited
buffers, and about the old LZ4 compressor doing the same.

It'd be nice to fix the logic that involves doing that if possible,
but for now, this at least reduces it to a small handful of messages
beyond whatever new code you're introducing.

### Description
Inited the biggest offendersin valgrind flooding output.

### How Has This Been Tested?
Ran it for a few days debugging other things, no problems found from it.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/zio.c [hunk 1]
```diff
@@ -1706,6 +1706,15 @@ zio_write_compress(zio_t *zio)
...
 	    !(zio->io_flags & ZIO_FLAG_RAW_COMPRESS)) {
 		void *cbuf = zio_buf_alloc(lsize);
+		/*
+		 * Both the crypto code and LZ4 code have bad habits of reading
+		 * uninited bytes sometimes.
+		 *
+		 * So to stop valgrind screaming to the point of uselessness,
+		 * zero them all.
+		 */
+		bzero(cbuf, lsize);
```
Comment (nabijaczleweli) [inline module/zfs/zio.c:1716]: As a preventative measure for rebase at a later date (#12996): `memset(cbuf, 0, lsize);`

Comment (rincebrain) [reply module/zfs/zio.c]: Done!

File: module/zfs/dsl_crypt.c [hunk 2]
```diff
@@ -1193,10 +1193,10 @@ dsl_crypto_key_sync(dsl_crypto_key_t *dck, dmu_tx_t *tx)
 {
 	zio_crypt_key_t *key = &dck->dck_key;
 	dsl_wrapping_key_t *wkey = dck->dck_wkey;
-	uint8_t keydata[MASTER_KEY_MAX_LEN];
-	uint8_t hmac_keydata[SHA512_HMAC_KEYLEN];
-	uint8_t iv[WRAPPING_IV_LEN];
-	uint8_t mac[WRAPPING_MAC_LEN];
+	uint8_t keydata[MASTER_KEY_MAX_LEN] = {0};
```
Comment (szubersk) [inline module/zfs/dsl_crypt.c:1196]: It is too early to have strong opinions since this PR is WIP, I must warn you though that `= {0}` generates `memset()`/`rep stos`. This might directly impact the performance, depends on how often this function is called in kernel space and how big the arrays are.

Comment (rincebrain) [reply module/zfs/dsl_crypt.c:1196]: I am aware that zeroing regions is usually not free, yes.

The claim, and reason for the PR, was that without fixing a number of existing uses in the codebase, trying to use things like `valgrind` just gets drowned out by a bunch of things referencing these regions.

If people think that's not worthwhile, that's fine, I'll make a note and avoid it in the future.

Comment (szubersk) [reply module/zfs/dsl_crypt.c:1196]: On the contrary, valgrind complains for a reason and memory corruption is one of the biggest plagues in C/C++ landscape. Fixing those errors is important. I just wanted to highlight potential unintended consequences.

**Comment by adamdmoss** [comment]
FWIW, the culprit in lz4 (and with a very similar trick in zstd, though not sure why valgrind hasn't noticed that) is almost certainly LZ4_wildCopy8() which is inlined and intentionally `can overwrite up to 8 bytes beyond dstEnd`.  It's not documented as such but the code appears as if it can over-read up to ~8 bytes also.  If the last part of the lz4 stream is literals then this is going to read <=8 bytes past the end of the lz4 data; I believe this is supposed to be safe by design for handwavey reasons (??!) but will leave valgrind correctly unhappy.
Similarly there is LZ4_wildCopy32() which can `overwrite up to 32 bytes beyond dstEnd`...

TLDR; zeroing just the last 32 bytes of `cbuf` should be similarly safe (for lz4) and faster.  It should be adequate for zstd too which more helpfully does at least `#define WILDCOPY_OVERLENGTH 32`

**Comment by adamdmoss** [comment]
Ah, I think perhaps I begin the grok the handwaving and why this affects reading undefined bytes from the *output* buffer rather than the *input* buffer - lz4/zstd decompression can copy literal bytes from the *output* stream to itself, and wildcopy will happily both read and write a bunch of undefined bytes if they overlap the end of the output cursor.
But since copying literals near the output cursor can happen at any point in the stream, not just at the end, the whole output buffer *should* be zero'd to ensure that an undefined bytes is never read. :(

**Comment by rincebrain** [comment]
I would suspect that there are cases where it's technically not guaranteed to be okay that valgrind is noticing and modern LZ4 (and zstd) probably have since caught.

It's also not the only culprit here.

**Comment by adamdmoss** [comment]
lz4 is written to intentionally and correctly operate that way but lz4 is also fast enough that I believe zero'ing the output buffer just to avoid valgrind errors is going to have a quite measurable impact on lz4 decompression speed.
Could the mitigation only be applied to DEBUG builds...?

**Comment by adamdmoss** [comment]
Ah and you're correct that our lz4 is so old that it might be an actual issue, not just a wildcopy trick. :(

**Comment by adamdmoss** [comment]
(Wouldn't it be cool if we could upgrade our LZ4 :grin: :grin:)

**Comment by rincebrain** [comment]
My bleeding edge LZ4 PR doesn't drop the old compressor, so that wouldn't save us.

To be frank, I'd rather close the PR than apply that change. If we think it's important things be initialized, do it. If we don't, don't. Don't just silence it where we have something complaining.

File: module/zfs/dsl_crypt.c (hunk 2 above)
Comment (mjguzik) [reply module/zfs/dsl_crypt.c:1196]: If this is file is shared by both kernel and userspace you can make zeroing conditional on compiling for the latter.

Are you sure the valgrind warning is a false-positive here?

Comment (behlendorf) [reply module/zfs/dsl_crypt.c:1196]: This function is only called when updating the key data, this should be a relatively infrequent operation and the function as implemented is already fairly expensive.  This change wont have any meaningful performance impact.  Why don't we at least more forward with this hunk which should be uncontroversial.

I agree, the `zio_write_compress()` change is likely too expensive and doing something reasonable there is going to be more difficult.

Comment (mjguzik) [reply module/zfs/dsl_crypt.c:1196]: So again, if this is a false positive or is otherwise safe to ignore for whatever reason, silencing valgrind may be hidden behind an ifdef (can't come up with a good name for it though). Then there is just some annoyance that you have to rebuild if you want to trace it, but given that other types of tooling would require rebuilding anyway to add instrumentation I don't think it's a big deal.

Perhaps ZFS_MEM_DEBUG?
